AMBARI-24714. Avoid multi-threading and JPA caching issues when aborting requests and processing agent reports at the same time

smolnar82 opened the pull request:
## What changes were proposed in this pull request?

User aborted task reported as FAILED in background operations:
1. Install cluster with all services ( Including Druid)
2. Click on Stop all services
3. When BGoperations window is loaded, click on the abort button for the same task
4. The operation should be aborted and marked as Aborted with "Aborted by user" message in the sub tasks logs
5. But instead the operation is marked as failure.

## How was this patch tested?

In addition to JUnit tests I used a little [Python script](https://issues.apache.org/jira/secure/attachment/12941947/abort_bg_operations.py) to reproduce the issue first (see in the attachment of the corresponding JIRA). Moreover I edited `ambari-agent.ini` and reduced command report/heartbeat periods. This was needed due to the nature of this issue.
When executed my tool (should abort 500 `start_all` - or `stop_all` - requests successfully) it failed very early (90% of the cases did not survive until the 30th round).
After I deployed the fix and executed the tool it succeeded.

smolnar82 committed:
AMBARI-24714. Avoid multi-threading and JPA caching issues when aborting requests and processing agent reports at the same time

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4134/
Test FAILed.
Test FAILured.

smolnar82 commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4135/
Test PASSed.

jonathan-hurley commented on ambari-server/src/main/java/org/apache/ambari/server/orm/dao/HostRoleCommandDAO.java:
```diff
@@ -286,30 +286,49 @@ public HostRoleCommandDAO(
-      return result;
+    if (refresh) {
+      for (HostRoleCommandEntity hostRoleCommand : result) {
+        entityManagerProvider.get().refresh(hostRoleCommand);
```
I'm worried about the refresh calls here. In large requests (such as those for upgrades), there could be 10's of 1000's of HRCs. Calling refresh on each one will incur DB hits and potentially cause the web client to timeout on things like Pause/Downgrade.

I think we need to do some more investigation of this code before it gets checked in. '

In general, I think that requiring Java-level locks around JPA means that something is wrong with our JPA design and that's where it should be fixed. `refresh()` is dangerous in that it makes it look like we're fixing a problem, yet we're introducing much larger ones that only customers see.

smolnar82 replied:
Thanks for your comment @jonathan-hurley 
I updated the code to avoid calling refresh within findByPKs; please review

smolnar82 replied:
Additional note: `ActionDBAccessor.updateHostRoleStates(Collection<CommandReport>) ` - where all this happens - is only invoked on the path of command report processing (from `HeartbeatProcessor`); it does not affect upgrades.

smolnar82 committed:
AMBARI-24724. Refreshing certain HRCs only to check their statuses again before we try to update them

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4136/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4137/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4149/
Test PASSed.

jonathan-hurley reviewed (commented):
Can you provide some more information about why this is needed? Calling `refresh()` on every HRC update as well as needing locks seems like we're dancing around a larger problem here...

dlysnichenko commented on ambari-server/src/main/java/org/apache/ambari/server/actionmanager/ActionDBAccessorImpl.java:
```diff
@@ -153,6 +155,8 @@
   private Cache<Long, HostRoleCommand> hostRoleCommandCache;
   private long cacheLimit; //may be exceeded to store tasks from one request
 
+  private final ReadWriteLock hrcOperationsLock = new ReentrantReadWriteLock();
```
It would be great to add some comment here describing the cases when this lock should be used. The variable name does not say too much

smolnar82 replied:
Added

smolnar82 committed:
AMBARI-24714. Get rid of refresh calls in HRCDao

smolnar82 commented:
The use case is:
- creating a START ALL (or STOP ALL) services request
- aborting it immediately

This is happening in Ambari:
- on THREAD 1 org.apache.ambari.server.actionmanager.ActionScheduler.doWork() processes the cancel event an aborts all the HRCs belong to the request in question
- asked Andrew O. if there was any change that we return a 'failed' event upon aborting a HRC on the agent side; he said that this works like this for a long time. SO something on the server side must have been changed. Within org.apache.ambari.server.agent.HeartbeatProcessor.processCommandReports(List<CommandReport>, String, Long) we interact with HRCs in the following cases on THREAD 2 (while processing command reports received from agents):

1. First we fetch the tasks map using org.apache.ambari.server.actionmanager.ActionDBAccessorImpl.getTasks(Collection<Long>). This code uses a Cache from Guava and this is the only place where we populate it (iterating over on HRCs fetched by hostRoleCommandDAO.findByPKs(List<Long>): in case the given HRC is not in the cahce we add it).

2. Second - on a FAILED report status - we check if the corresponding HRC is in progress or not using org.apache.ambari.server.actionmanager.ActionDBAccessorImpl.getTask(long, boolean)

3. Finally we update state machines from reports using actionManager.processTaskResponse(hostName, reports, commands) (see at the very end of the method)


During my tests I found that any of these operations on THREAD 2 has been executed in the middle of THREAD 1 work thus I introduced locking: while we are aborting HRCs on THREAD 1 we should not allow reading/updating HRC states on THREAD 2. I expected that this solves the issue since the persisten context should be in good shape (all ABORTED HRCs are merged; I also modified this to merge one by one before we invoke audit log since it was not correct to log as "ABORTED" when we simply iterated over and populated fields but actual merging happend later).
Unfortunately it was not enough; still fetched `old` data from the persistence context; that's why I added the refresh calls.

**But...**

During the time I was writing this explanation I noticed the following:
- org.apache.ambari.server.actionmanager.ActionDBAccessorImpl.abortOperation(long) did not populate the Guava Cache which - IMO - should be done
- I missed a Java lock in org.apache.ambari.server.actionmanager.ActionDBAccessorImpl.getTasks(Collection<Long>) whereas I should have added it since it fetches HRCs from the persistent context

After implementing these two items and removed the 'refresh' calls it was all good.

Please review my latest patch.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4162/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4163/
Test PASSed.

jonathan-hurley reviewed (approved):
Awesome! Thanks for the in-depth explanation here. Glad that you were able to get rid of the refresh calls and deal only with the thread locks and cache updates.
